{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import permutations\n",
    "import numpy as np\n",
    "import copy\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "from pandas import DataFrame \n",
    "import os\n",
    "from math import pi\n",
    "\n",
    "import bokeh\n",
    "from bokeh.io import output_file, show\n",
    "from bokeh.models import BasicTicker, ColorBar, LinearColorMapper, ColumnDataSource, FixedTicker, PrintfTickFormatter\n",
    "from bokeh.plotting import figure\n",
    "from bokeh.transform import transform\n",
    "from bokeh.layouts import gridplot\n",
    "\n",
    "from bokeh.models import (BoxZoomTool, Circle, HoverTool,\n",
    "                          MultiLine, Plot, Range1d, ResetTool,)\n",
    "from bokeh.palettes import Spectral4\n",
    "from bokeh.palettes import Viridis256\n",
    "from bokeh.plotting import from_networkx\n",
    "\n",
    "import networkx as nx\n",
    "import community\n",
    "\n",
    "\n",
    "\n",
    "import colorcet\n",
    "cmap = colorcet.glasbey_dark\n",
    "\n",
    "\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "\n",
    "bokeh.io.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def needleman_wunsch(seq1, seq2, gap):\n",
    "    '''\n",
    "    Code to determine the needleman wunsch distance between two sequences, using the bolsum_50 matrix.\n",
    "    '''\n",
    "    amino_acid_dict = {'A':0, 'R':1, 'N':2, 'D':3, 'C':4, 'Q':5, 'E':6, 'G':7, 'H':8, 'I':9,  'L':10, 'K':11, 'M':12, 'F':13, 'P':14, 'S':15, 'T':16, 'W':17, 'Y':18, 'V':19}\n",
    "    \n",
    "    blosum50 = [[ 5,-2, -1, -2, -1, -1, -1,  0, -2, -1, -2, -1, -1, -3, -1,  1,  0, -3, -2,  0, -2],\n",
    "            [-2, 7, -1, -2, -4,  1,  0, -3,  0, -4, -3,  3, -2, -3, -3, -1, -1, -3, -1, -3, -1],\n",
    "            [-1,-1,  7,  2, -2,  0,  0,  0,  1, -3, -4,  0, -2, -4, -2,  1,  0, -4, -2, -3,  5],\n",
    "            [-2,-2,  2,  8, -4,  0,  2, -1, -1, -4, -4, -1, -4, -5, -1,  0, -1, -5, -3, -4,  6],\n",
    "            [-1,-4, -2, -4, 13, -3, -3, -3, -3, -2, -2, -3, -2, -2, -4, -1, -1, -5, -3, -1, -3],\n",
    "            [-1, 1,  0,  0, -3,  7,  2, -2,  1, -3, -2,  2,  0, -4, -1,  0, -1, -1, -1, -3,  0],\n",
    "            [-1, 0,  0,  2, -3,  2,  6, -3,  0, -4, -3,  1, -2, -3, -1, -1, -1, -3, -2, -3,  1],\n",
    "            [ 0,-3,  0, -1, -3, -2, -3,  8, -2, -4, -4, -2, -3, -4, -2,  0, -2, -3, -3, -4, -1],\n",
    "            [-2, 0,  1, -1, -3,  1,  0, -2, 10, -4, -3,  0, -1, -1, -2, -1, -2, -3,  2, -4,  0],\n",
    "            [-1,-4, -3, -4, -2, -3, -4, -4, -4,  5,  2, -3,  2,  0, -3, -3, -1, -3, -1,  4, -4],\n",
    "            [-2,-3, -4, -4, -2, -2, -3, -4, -3,  2,  5, -3,  3,  1, -4, -3, -1, -2, -1,  1, -4],\n",
    "            [-1, 3,  0, -1, -3,  2,  1, -2,  0, -3, -3,  6, -2, -4, -1,  0, -1, -3, -2, -3,  0],\n",
    "            [-1,-2, -2, -4, -2,  0, -2, -3, -1,  2,  3, -2,  7,  0, -3, -2, -1, -1,  0,  1, -3],\n",
    "            [-3,-3, -4, -5, -2, -4, -3, -4, -1,  0,  1, -4,  0,  8, -4, -3, -2,  1,  4, -1, -4],\n",
    "            [-1,-3, -2, -1, -4, -1, -1, -2, -2, -3, -4, -1, -3, -4, 10, -1, -1, -4, -3, -3, -2],\n",
    "            [ 1,-1,  1,  0, -1,  0, -1,  0, -1, -3, -3,  0, -2, -3, -1,  5,  2, -4, -2, -2,  0],\n",
    "            [ 0,-1,  0, -1, -1, -1, -1, -2, -2, -1, -1, -1, -1, -2, -1,  2,  5, -3, -2,  0,  0],\n",
    "            [-3,-3, -4, -5, -5, -1, -3, -3, -3, -3, -2, -3, -1,  1, -4, -4, -3, 15,  2, -3, -5],\n",
    "            [-2,-1, -2, -3, -3, -1, -2, -3,  2, -1, -1, -2,  0,  4, -3, -2, -2,  2,  8, -1, -3],\n",
    "            [ 0,-3, -3, -4, -1, -3, -3, -4, -4,  4,  1, -3,  1, -1, -3, -2,  0, -3, -1,  5, -3]]\n",
    "    \n",
    "    # Create an alignment matrix to store match and mismatch values\n",
    "    alignment_matrix = np.zeros((len(seq1) + 1, len(seq2) + 1))\n",
    "        \n",
    "    # Create an position matrix to store what positions the current value came from\n",
    "    position_matrix = np.zeros((len(seq1) + 1, len(seq2) + 1))\n",
    "    \n",
    "    # Set the boundary conditions for the matrix\n",
    "    for i in range(len(seq1) + 1):\n",
    "        alignment_matrix[i, 0] += i*gap\n",
    "    \n",
    "    for j in range(len(seq2) + 1):\n",
    "        alignment_matrix[0, j] += j*gap\n",
    "        if j > 0:\n",
    "            position_matrix[0, j] += 1\n",
    "    \n",
    "    # Loop through the matrix\n",
    "    for i in range(1, len(seq1) + 1):\n",
    "        for j in range(1, len(seq2) + 1):            \n",
    "            value = blosum50[amino_acid_dict[seq1[i-1]]][amino_acid_dict[seq2[j-1]]]\n",
    "                \n",
    "            match_list = [alignment_matrix[i-1, j] + gap, alignment_matrix[i, j-1] + gap, alignment_matrix[i-1, j-1] + value]\n",
    "            \n",
    "            inds = np.argmax(match_list)\n",
    "            \n",
    "            alignment_matrix[i, j] += match_list[inds] \n",
    "        \n",
    "            position_matrix[i, j] = inds\n",
    "    \n",
    "    seq1_alignment = []\n",
    "    seq2_alignment = []\n",
    "    \n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    \n",
    "    while (n or m) > 0:\n",
    "        val = position_matrix[n,m]\n",
    "    \n",
    "        if val == 2:\n",
    "            seq1_alignment = [seq1[n-1]] + seq1_alignment \n",
    "            seq2_alignment = [seq2[m-1]] + seq2_alignment\n",
    "            \n",
    "            n = n - 1\n",
    "            m = m - 1\n",
    "            \n",
    "        elif val == 0:\n",
    "            seq1_alignment = [seq1[n-1]] + seq1_alignment \n",
    "            seq2_alignment = ['_'] + seq2_alignment\n",
    "            n -= 1\n",
    "            \n",
    "        elif val == 1:\n",
    "            seq1_alignment = ['_'] + seq1_alignment\n",
    "            seq2_alignment = [seq2[m-1]] + seq2_alignment\n",
    "            \n",
    "        \n",
    "            m -= 1\n",
    "    \n",
    "    seq1_str = ''.join(seq1_alignment)\n",
    "    seq2_str = ''.join(seq2_alignment)\n",
    "        \n",
    "    return(seq1_str, seq2_str, alignment_matrix[-1, -1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Edit to be your filepath\n",
    "fname = 'Priya_R2_DNAandvirus_over10_Tekbrain_over50_nodup_nospikein_B.csv'\n",
    "\n",
    "# Could convert to csv if needed\n",
    "df = pd.read_csv(fname, comment = '#')\n",
    "\n",
    "#top50 variants\n",
    "df = df.nlargest(50, 'Tek-mean-synth')\n",
    "\n",
    "#Alternatively, you could select ositively enriched variants\n",
    "#df = df.loc[(df['Tek-mean-synth']>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the number of variants you have in your dataset, scales NxN so should be kept small.\n",
    "n = len(df)\n",
    "\n",
    "alignment_scores = np.empty((n, n))\n",
    "alignment_scores.fill(np.nan)\n",
    "\n",
    "# There is some nuance here... The [2:-2] is to remove amino acids that are invariable from the sequences. This may need to be adjusted based on the data you're working with. -8 is the gap penalty, which allows amino acids that are in different locations to be aligned with a penalty\n",
    "for i in tqdm(range(n)):\n",
    "    alignment_scores[i,i] = needleman_wunsch(df['Amino Acid'][i][2:-2],df['Amino Acid'][i][2:-2],-8)[2]\n",
    "    \n",
    "    for j in range(i, n):\n",
    "        alignment_scores[i,j] = (needleman_wunsch(df['Amino Acid'][i][2:-2],df['Amino Acid'][j][2:-2],-8)[2])\n",
    "         \n",
    "        alignment_scores[j,i] =  alignment_scores[i,j] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dataframes for normalized scores\n",
    "normalized_alignment_scores = np.empty((n, n))\n",
    "normalized_alignment_scores.fill(np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize the scores\n",
    "for i in range(n):\n",
    "    normalized_alignment_scores[i] = (alignment_scores[i] - np.nanmin(alignment_scores[i])) / (alignment_scores[i,i] - np.nanmin(alignment_scores[i]))\n",
    "\n",
    "normalized_alignment_scores[-1, -1] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe for plotting\n",
    "normalized_alignment_df = pd.DataFrame(\n",
    "    normalized_alignment_scores,\n",
    "    columns=df['Amino Acid'][0:len(df)],\n",
    "    index=df['Amino Acid'][0:len(df)])\n",
    "normalized_alignment_df.index.name = 'source'\n",
    "normalized_alignment_df.columns.name = 'target'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data.frame in the right format\n",
    "stacked_normalized_alignment_df = normalized_alignment_df.stack().rename(\"value\").reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot the data\n",
    "\n",
    "mapper = LinearColorMapper(\n",
    "    palette='Viridis256', low=stacked_normalized_alignment_df['value'].min(), high=stacked_normalized_alignment_df['value'].max())\n",
    "\n",
    "# Define a figureabsabs\n",
    "p = figure(\n",
    "    plot_width=1000,\n",
    "    plot_height=1000,\n",
    "    x_range=list(stacked_normalized_alignment_df['source'].drop_duplicates()),\n",
    "    y_range=list(stacked_normalized_alignment_df['target'].drop_duplicates()),\n",
    "    title=\"Positively enriched variants from CNS (mouse)\",\n",
    "    toolbar_location='right',\n",
    "    tools = \"pan,wheel_zoom,box_zoom,reset,save\",\n",
    "    x_axis_location=\"above\")\n",
    "\n",
    "# Create rectangle for heatmap\n",
    "p.rect(\n",
    "    x=\"source\",\n",
    "    y=\"target\",\n",
    "    width=1,\n",
    "    height=1,\n",
    "    source=ColumnDataSource(stacked_normalized_alignment_df),\n",
    "    line_color=None,\n",
    "    fill_color=transform('value', mapper))\n",
    "\n",
    "# Add legend\n",
    "color_bar = ColorBar(\n",
    "    color_mapper=mapper,\n",
    "    location=(0, 0))\n",
    "\n",
    "p.add_layout(color_bar, 'right')\n",
    "p.y_range.range_padding = 0.005\n",
    "p.x_range.range_padding = 0.005\n",
    "\n",
    "p.xaxis.major_label_orientation = pi/4\n",
    "\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbrs = NearestNeighbors(n_neighbors=10, algorithm='ball_tree').fit(normalized_alignment_scores)\n",
    "distances, indices = nbrs.kneighbors(normalized_alignment_scores)\n",
    "\n",
    "A = nbrs.kneighbors_graph(normalized_alignment_scores).toarray()\n",
    "\n",
    "G = nx.from_numpy_matrix(A)\n",
    "partition = community.best_partition(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#version without the labels\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "matplotlib.rcParams['figure.figsize'] = 12,8\n",
    "\n",
    "pos = nx.spring_layout(G, k=0.01)\n",
    "\n",
    "communities = []\n",
    "\n",
    "for i, com in enumerate(set(partition.values())):\n",
    "    list_nodes = [nodes for nodes in partition.keys()\n",
    "                                if partition[nodes] == com]\n",
    "    \n",
    "    #print(df['Amino Acid'][list_nodes][:])\n",
    "    nx.draw_networkx_nodes(G, \n",
    "                           pos, \n",
    "                           list_nodes,\n",
    "                           node_size=200,\n",
    "                           node_color=cmap[i])\n",
    "\n",
    "nx.draw_networkx_labels(G, pos, labels = df['Rank'], font_size=12, font_color='w', font_weight='bold')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
